mesonbuild/compilers/compilers.py

```diff
@@ -503,6 +503,22 @@ def get_base_link_args(options, linker, is_shared_module):
         pass
     return args
 
+def prepare_rpaths(raw_rpaths, build_dir, from_dir):
+    internal_format_rpaths = [evaluate_rpath(p, build_dir, from_dir) for p in raw_rpaths]
+    ordered_rpaths = order_rpaths(internal_format_rpaths)
+    return ordered_rpaths
+
+def order_rpaths(rpath_list):
+    # We want rpaths that point inside our build dir to always override
+    # those pointing to other places in the file system. This is so built
+    # binaries prefer out libraries to the ones that may lie somewhere
+    # in the file system, such as /lib/x86_64-linux-gnu.
+    #
+    # The correct thing to do here would be C++'s std::stable_partition.
+    # Python standard library does not have it, so replicate it.
+    return [x for x in rpath_list if not os.path.isabs(x)] + \
+        [x for x in rpath_list if os.path.isabs(x)]
+
```

Comment:
I think you can simplify that even more to:
```python
return sorted(rpath_list, key=os.path.isabs)
```

Reply:
Sure! For some reason I was thinking that `rpath_list` needs to be sorted alphabetically.

Reply:
I don't think so, I think we just need to put meson internal paths before system ones.

Reply:
It would not make much sense, indeed, and it is not what @jpakkane  code does, anyway.

Reply:
Updated.

Reply:
Looks good, but the comment above the code seems off. I'm am not familiar with what `std::stable_partition` does but the comment makes it look like the Python `sorted` solution is inferior, and I don't think it is. It is just a different syntax to express the same.

Reply:
I think the comment was correct for the original code, python's `sorted` is basically a more powerful version of `std::stable_partition`

Reply:
Or, alternatively, `std::stable_partition` is a more specific name for the operation. You immediately know what it does without needing to read any source code.

Reply:
The behaviour should be identical in this particular case. `False` is less than `True` and items with the same value retain their order.

Reply:
yup. Just in general, they're a little different.

Reply:
I find it strange to explain what a piece of Python code in a Python project does in term of the C++ standard library (even more so in a way that hints to the fact that the Python solution is inferior). But maybe it is just me, and the fact that I'm more familiar with Python than with C++.

Reply:
It's not a question of inferiorness. It's just that the C++ standard library has a function that would do _exactly_ what we need. It is mentioned mostly so that people can refer to existing documentation as needed.

Reply:
Sorry for carrying on the academic discussion, I have too many thing I'm procrastinating :) The Python documentation for `sorted` fulfills the same goal, and it is a much more natural resource for someone reading Python code.